File: LanguageWrappers/python/report.py
import tempfile
import os
import shutil
import subprocess



class Report :
	"""
	Report Class
	This class provides functionality to generate and export reports using the RDL (Report Definition Language) command-line tool. 
	It supports exporting reports in various formats such as PDF, CSV, XLSX, XML, RTF, TIF, and HTML.
	Usage:
	1. Copy and paste the `report.py` file into your Python project.
	2. Import the `Report` class into your script:
		```python
		import report
		```
	3. Create an instance of the `Report` class:
		```python
		rpt = report.Report(report_path="path_to_report.rdl", rdl_cmd_path="path_to_rdl_cmd.exe", path_to_dotnet="path_to_dotnet_executable")
		```
		- `report_path`: Path to the RDL file.
		- `rdl_cmd_path`: Path to the RDL command-line executable.
		- `path_to_dotnet`: (Optional) Path to the .NET executable, if required.
	4. Set report parameters (if any):
		```python
		rpt.set_parameter("ParameterName", "ParameterValue")
		```
	5. Set the connection string (if required):
		```python
		rpt.set_connection_string("your_connection_string")
		```
	6. Export the report to a file:
		```python
		rpt.export(type="pdf", export_path="output_path.pdf")
		```
		- `type`: The export format. Supported formats are "pdf", "csv", "xslx", "xml", "rtf", "tif", "html". Defaults to "pdf".
		- `export_path`: The path where the exported file will be saved.
	7. Export the report to memory (for direct display):
		```python
		data = rpt.export_to_memory(type="pdf")
		```
		- `type`: The export format. Supported formats are "pdf", "csv", "xslx", "xml", "rtf", "tif", "html". Defaults to "pdf".
		- Returns the exported report data as a string or binary, depending on the format.
	"""
	__report_path=""
	__parameters = {}
	__connection_string = None
	__rdl_cmd_path=""
	__path_to_dotnet = None
	
	def __init__(self, report_path: str, rdl_cmd_path: str, path_to_dotnet :str =None):
		self.__report_path = report_path
		self.__rdl_cmd_path = rdl_cmd_path
		self.__path_to_dotnet = path_to_dotnet
	
	def set_parameter(self, name, value):
		'''
		Set a parameters values
		 name - string - the report parameter name
		 value - string - the value of the parameter
		'''
		
		self.__parameters.update({name:value})
		
	def set_connection_string(self, connection_string: str):
		'''
		Set the connection string for the report
		 connection_string - string - the connection string to use for the report
		'''
		
		self.__connection_string = connection_string


	def export(self, type : str, export_path : str):
		'''
		 Export report to a file on the server
		 type - string - Export type "pdf", "csv", "xslx", "xml", "rtf", "tif", "html".  If type does not match it will default to PDF.
		 export_type - string - path on server to export file
		'''
		if (type != "pdf" and type != "csv" and type != "xslx" and type != "xml" and type != "rtf" and type != "tif" and type != "html"):
			type = "pdf"
			
		cmd = []
		if (self.__path_to_dotnet != None):
			cmd.append(self.__path_to_dotnet)

		cmd.append(self.__rdl_cmd_path)
	
		fd, temp_name = tempfile.mkstemp()
		os.close(fd)
		temp_folder = os.path.dirname(temp_name)
		shutil.copyfile(self.__report_path, temp_name)
		
		#add path to rdl file
		rdl_path = "/f" + temp_name
		
		# Add all parameters to report
		count=0
		for key in self.__parameters:
			if (count == 0):
				rdl_path = rdl_path + '?' + key + '=' + self.__parameters[key]
			else:
				rdl_path = rdl_path + '&' + key + '=' + self.__parameters[key]
		
			count = count + 1
		
		cmd.append(rdl_path)
		
		# set the export type
		cmd.append('/t' + type)
		
		#set the folder that the file will be exported
		cmd.append('/o' + temp_folder)

		if self.__connection_string:
			cmd.append('/c' + self.__connection_string+'')
		
		#print("Executing command:", cmd)

		subprocess.run(cmd)

		temp_pdf = temp_folder + os.sep + os.path.basename(temp_name) + "." + type
		final_pdf = export_path

		shutil.copyfile(temp_pdf, final_pdf)
		os.remove(temp_name)
		os.remove(temp_pdf)
	


	def export_to_memory(self, type : str) -> bytes | str :
		'''
		Export report to memory for direct display on page
		type - string - Export type "pdf", "csv", "xslx", "xml", "rtf", "tif", "html".  If type does not match it will default to PDF.
		'''
		if (type != "pdf" and type != "csv" and type != "xslx" and type != "xml" and type != "rtf" and type != "tif" and type != "html"):
			type = "pdf"
		
		fd, temp_name = tempfile.mkstemp()
		os.close(fd);
		self.export(type, temp_name)
		if type == "pdf" or type == "tif" or type == "rtf" or type == "xslx":
			with open(temp_name, 'rb') as binary_file:
				data = binary_file.read()
		else:
			with open(temp_name, 'r+') as text_file:
				data = text_file.read()

		os.remove(temp_name)
	
		return data
	
	

File: LanguageWrappers/ruby/report.rb
require 'tempfile'

class Report

	def initialize(report_path, rdl_cmd_path, path_to_dotnet = nil)
		@@report_path = report_path
		@@path_to_rdlcmd = rdl_cmd_path
		@@path_to_dotnet = path_to_dotnet
		@@parameters = {}
	end
	
	def set_parameter(name, value)
		#Set a parameters values
		# name - string - the report parameter name
		# value - string - the value of the parameter
		
		@@parameters[name] = value
		
	end

	def set_connection_string(connection_string)
		#Set the connection string for the report
		# connection_string - string - the connection string to use for the report
		
		@@connection_string = connection_string
	end

	def export(type, export_path)
		# Export report to a file on the server
		# type - string - Export type "pdf", "csv", "xslx", "xml", "rtf", "tif", "html".  If type does not match it will default to PDF.
		# export_type - string - path on server to export file

		if (type != "pdf" and type != "csv" and type != "xslx" and type != "xml" and type != "rtf" and type != "tif" and type != "html")
			type = "pdf"
		end
			
		cmd = []
		if @@path_to_dotnet
			cmd.push(@@path_to_dotnet)
		end

		cmd.push(@@path_to_rdlcmd)
		
		temp_file_to_close = Tempfile.new('maj')
		temp_name = temp_file_to_close.path
		temp_file_to_close.close
		temp_folder = File.dirname(temp_name)
		FileUtils.cp(@@report_path, temp_name)
		
		#add path to rdl file
		rdl_path = "/f" + temp_name
		
		# Add all parameters to report
		count=0
		
		@@parameters.each do |key, value|
			if (count == 0)
				rdl_path = rdl_path + '?' + key + '=' + value
			else
				rdl_path = rdl_path + '&' + key + '=' + value
			end
		
			count = count + 1
		end
		
		cmd.push(rdl_path)
		
		# set the export type
		cmd.push('/t' + type)
		
		#set the folder that the file will be exported
		cmd.push('/o' + temp_folder)

		if @@connection_string
			cmd.push('/c' + @@connection_string)
		end

		#puts cmd.join(" ")
		
		IO.popen(cmd) do |io|
			 # wait until process finished
			io.readlines
		end 
		
		temp_pdf = temp_folder + File::SEPARATOR + File.basename(temp_name) + "." + type
		final_pdf = export_path
		#echo(cmd)
		FileUtils.cp(temp_pdf, final_pdf)
		
		File.delete(temp_name)
		File.delete(temp_pdf)
	
	end


	def export_to_memory(type)
		# Export report to memory for direct display on page
		# type - string - Export type "pdf", "csv", "xslx", "xml", "rtf", "tif", "html". If type does not match it will default to PDF.

		if !["pdf", "csv", "xslx", "xml", "rtf", "tif", "html"].include?(type)
			type = "pdf"
		end

		temp_file = Tempfile.new('export')
		temp_name = temp_file.path
		temp_file.close
		temp_file.unlink

		export(type, temp_name)

		data = nil
		if ["pdf", "tif", "rtf", "xslx"].include?(type)
			File.open(temp_name, 'rb') do |binary_file|
				data = binary_file.read
			end
		else
			File.open(temp_name, 'r') do |text_file|
				data = text_file.read
			end
		end

		File.delete(temp_name) if File.exist?(temp_name)

		data
	end
	
end

